doc/design/error_handling_requirements.md

```diff
@@ -0,0 +1,228 @@
+# Error Handling
+
+## Faults, Errors and Failures
+
+We use the following error terminology.
+
+1. A **fault** is a design flaw, a programming error or can e.g. be caused by unforseen use
+   (oversight of a use case is basically also a design flaw).
+2. An **error** is the result of a fault and manifests as an incorrect or inconsistent (intermediate) result.
+   An error may not be externally observable by the user of the system.
+3. A **failure** occurs if the system observably acts in contradiction to its specification.
+
+This reasonably assumes that errors are always caused by faults. If these errors are not detected
+they can cause failures. By detecting and handling errors failures are avoided.
+
+The goal of this document is to define error handling requirements.
+
+## Error Detection vs. Error Handling
+
+To handle an error it obviously needs to be detected first. To do so, the code must contain conditionals that check
+for inconsistent states, unavailability of resources, return codes of system calls etc.
+This means that the error has to be anticipated by the system designer,
+i.e. we cannot handle unforseen errors since they are not detected by definition.
+
+Error handling is the execution of handling code once the error was detected.
+This ranges from simple logging to sophisticated recovery or shutdown code and depends on the error in question.
+We distinguish between recoverable and non-recoverable error handling.
+
+### Recoverable Error Handling
```

Comment:
As a thought to the recoverable vs. non-recoverable error handling:
Should we maybe state that when a error is non recoverable in iceoryx we always call the error handler and the error handler has only one error level fatal? This would simplify a lot and would provide us with a clean rule.
All recoverable errors always use `expected` for error handling (maybe we should rename it to result soon).

A non recoverable error in iceoryx would be something for me like:
* accessing non existing memory like, `optional.value()` without having a value or `vector[i], we can decide for undefined behavior or segfault or call the error handler`. In the end non recoverable means in the context of iceoryx that the next line will either lead to undefined behavior, a segfault or some other crash.

This can maybe can be mitigated so that this becomes a recoverable error, for instance by using exceptions in the error handler but on our level this is non recoverable.

A recoverable error would be something where, caused by this failure, at no time undefined behavior or a crash is immanent from our point of view. Something like creating shared memory but the user does not have the permission.

Reply:
@mossmaurice the user must be able to override bode, the error handling and the error reporting.

AFAIK @MatthiasKillat is moving toward like

```cpp
if (someCall.has_error())
{
    IOX_RAISE(error level, error code) << "Be careful an recoverable error happened!";
    return cxx::error<BlubbError>(BlubbError::SOME_ERROR);
}
```

So one line less. I would also decouple the `cxx::error` from the error reporting/handling. Since hoofs is a library, we should not force our users to a specific pattern and the expected is totally fine to use without the error reporter/handler. Additionally, not every time we use the `expected` it is warranted to also report the error. Sometimes it might be better to bubble up the error until enough context is available. For example, we use the `expected` for `Subscriber::takeChunk` and when there is no chunk available we do not want to report this since this is not an error when we poll but we communicate it with the `cxx::error`.

Reply:
@elBoberido 
> the user must be able to override bode, the error handling and the error reporting.

For sure, that's what I meant. Shame on me if you understood it differently :sweat_smile: 

> Since hoofs is a library, we should not force our users to a specific pattern and the expected is totally fine to use without the error reporter/handler. 

Good point, that should definitely not happen. The "how" can be figured out for sure somehow (inheritance, special mode, etc.). My main fear is that reporting a recoverable error will be forgotten both when writing the code and when reviewing it. IMHO it is error-prone, just look at how chaotic it is in the current code. I would prefer to have concise rules like A) recoverable don't need to call `IOX_RAISE` B) non-recoverable error means call `IOX_RAISE` (see examples depicted below). Also, why do I have to show my intention of reporting an error to the user two times? For me the idiomatic way is the return type `cxx::error`.

> Additionally, not every time we use the expected it is warranted to also report the error.

I agreed with the `takeChunk` case, although that is already some sort of a misuse of `std::expected`. However, an error is still an error. I would feel much safer preemptively reporting all of them after reading [this book](https://www.amazon.de/-/en/Chris-Hobbs/dp/0367338858/ref=sr_1_1?crid=XVDHEP7F0YE1&keywords=chris+hobbs&qid=1656075258&sprefix=chris+hobbs%2Caps%2C97&sr=8-1).


I couldn't yet answer the question to external contributors if someone would ask: When should I use `IOX_RAISE` in the code? What is the purpose of it?

Summed up my two suggestions are:

1. Clear and concise rules for recoverable and non-recoverable errors (like @elfenpiff suggested)
2. Stronger differentiate between error reporting and error handling to the user (Who handles the error? Me, the API caller, the system above or god?)

## Recoverable errors

```

std::expected<FooBar, FooBarError> doSomethingHighLevel()
{

   if(doSomethingLowLevel().has_error())
   {
      // IMHO the next line is redundant and can be error-prone
      IOX_RAISE(error level, error code) << "Be careful an recoverable error happened!"; // better IOX_REPORT
      // Handle or pass up with automatic conversion by implementing the mapping as discussed before
      return cxx::error<FooBarError>(doSomethingLowLevel().get_error());
   }

}

std::expected<FuuBaz, FuuBazError> doSomethingLowLevel()
{
   if (someThing.has_error())
   {
       // IMHO the next line is redundant and can be error-prone
       IOX_RAISE(error level, error code) << "Be careful an recoverable error happened!"; // better IOX_REPORT
       return cxx::error<BlubbError>(BlubbError::SOME_ERROR);
   }
}

```

My wish would be to see the error callstack or propoagation of the errors through the different layers in the fault recorder.

If we stick with adding the `IOX_RAISE` manually, I still would like to rename it to something like `IOX_REPORT`. We are just reporting and error here and are not handling it. The error being handled by the user via `cxx::error` :exclamation:  Of course `IOX_RAISE` can be called underneath non-fatally.

## Non-recoverable errors

```
T& vector::operator[](uint64_t index)
{
    // This calls IOX_RAISE(error level, error code) underneath
    IOX_ASSERT(index < m_size) << "Out of bounds access";
    return reinterpret_cast<const T*>(m_data)[index];
}
```

@MatthiasKillat Could you add such specific examples for both use cases? Thanks!

Reply:
@elBoberido Sure, on macro `IOX_RAISE` is fine. I also agree with the error handler being the lowest building block. However, my point of the above post was mainly to not burden developers with the necessity to add `IOX_RAISE` for recoverable errors. IMHO we should report them silently underneath. For non-recoverable errors I'm fully for adding an explicit `IOX_RAISE`!